### PR TITLE
Assign phandles to all labelled nodes if -@ is specified to match GPL dtc(1) behavior

### PR DESCRIPTION
**Pull request by kevans91**
libfdt assumes that all nodes in a symbol'd tree will have phandles assigned to them, as this is how GPL dtc(1) operates. They're not entirely flexible on changing this, so modify our dtc(1) to match their behavior.

### CODE REVIEW

File: fdt.cc [hunk 1]
```diff
@@ -1730,7 +1747,13 @@ device_tree::parse_dts(const string &fn, FILE *depfile)
 		}
 	}
 	collect_names();
-	resolve_cross_references();
+	uint32_t phandle = 1;
+	// If we're writing symbols, go ahead and assign phandles to the entire
+	// tree. We'll do this before we resolve cross references, just to keep
+	// order semi-predictable and stable.
+	if (write_symbols)
```
Comment (davidchisnall) [inline fdt.cc]: Please add braces around the block.

File: fdt.cc [hunk 2]
```diff
@@ -1230,8 +1230,62 @@ device_tree::collect_names()
...
+		}
+	}
+
+	return (p);
+}
+
+void
+device_tree::assign_phandles(node_ptr &n, uint32_t &next)
+{
+	if (!n->labels.empty())
```
Comment (davidchisnall) [inline fdt.cc]: Please put braces around if statement bodies.

File: fdt.cc [hunk 3]
```diff
@@ -1230,8 +1230,62 @@ device_tree::collect_names()
...
+		assign_phandle(n.get(), next);
+
+	for (auto &c : n->child_nodes())
```
Comment (davidchisnall) [inline fdt.cc]: And around loop bodies.

File: fdt.hh [hunk 4]
```diff
@@ -799,6 +799,10 @@ class device_tree
 	 * properties that have been explicitly added.  
 	 */
 	void collect_names_recursive(node_ptr &n, node_path &path);
+	/**
+	 * Assign a phandle property to a single node.
```
Comment (davidchisnall) [inline fdt.hh]: Please document what this does with the `next` parameter.

File: fdt.hh [hunk 5]
```diff
@@ -814,7 +818,7 @@ class device_tree
 	 * node must have their values replaced by either the node path or
 	 * phandle value.
 	 */
-	void resolve_cross_references();
+	void resolve_cross_references(uint32_t &phandle);
```
Comment (davidchisnall) [inline fdt.hh]: Please document what this does with the `phandle` parameter.

File: fdt.cc (hunk 1 above)
Comment (kevans91) [reply fdt.cc]: Heh, sorry, habits. =)

**Pull request merged**